async-signature/src/lib.rs

```diff
@@ -33,7 +34,6 @@ pub trait AsyncSigner<S: 'static> {
     async fn sign_async(&self, msg: &[u8]) -> Result<S, Error>;
 }
 
-#[async_trait(?Send)]
 impl<S, T> AsyncSigner<S> for T
```

Comment:
I'd much rather have a `impl<D, S, T> AsyncSigner<S> for T where T: AsyncDigestSigner<D, S>` instead of this blanket.

Reply:
First, you cannot write such a blanket impl, because `D` is unconstrained (and indeed working around that issue is the whole purpose of the [signature-derive](https://github.com/RustCrypto/traits/blob/master/signature_derive/src/lib.rs):

```
error[E0207]: the type parameter `D` is not constrained by the impl trait, self type, or predicates
```

(And though `PrehashSigner` can define a default digest, you still won't be able to write a valid blanket impl with it)

Second, these traits are intended to correspond 1:1 with the traits in `signature`. If there isn't a blanket impl between `Signer` and `DigestSigner`, it doesn't make sense to have one here because that would be inconsistent.

Third, these blanket impls exist so that you can use a `Signer` as an `AsyncSigner`, i.e. if you have an API that accepts `AsyncSigner`, anyone can plug in a `Signer`, and the same goes for all of the other traits, eliminating the need to duplicate APIs everywhere just for async.

Reply:
I oversimplified the previous comment a bit, but yeah I intended:
``` rust
impl<S, T> AsyncSigner<S> for T
where
    S: PrehashSignature + 'static,
    T: AsyncDigestSigner<S::Digest, S>,
{
    async fn sign_async(&self, msg: &[u8]) -> Result<S, Error> {
        self.sign_digest_async(S::Digest::new_with_prefix(msg))
            .await
    }
}
```

This is essentially what's in the `#[derive(signature::Signer)]` I guess I could add that to `signature_derive` then?

Reply:
That does indeed seem to work, but `signature` has no such blanket impl, so I think it'd be weird for them to be inconsistent.

It could potentially be included in a hypothetical future `signature` 3.0, although the devil is in the details for these sort of blanket impls, and they may preclude overlapping impls with a valid use case.